pkg/proxy/backend/authenticator.go

```diff
@@ -24,74 +24,123 @@ import (
-			return writeErr
+
+	proxyCapability := auth.supportedServerCapabilities
+	if frontendTLSConfig == nil {
+		proxyCapability ^= pnet.ClientSSL
+	}
+	if err := clientIO.WriteInitialHandshake(proxyCapability.Uint32(), salt, mysql.AuthCachingSha2Password); err != nil {
+		return errors.WithStack(err)
+	}
+	pkt, isSSL, err := clientIO.ReadSSLRequestOrHandshakeResp()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	frontendCapability := pnet.Capability(binary.LittleEndian.Uint32(pkt))
+	if isSSL {
+		if proxyCapability&pnet.ClientSSL == 0 {
+			return clientIO.WriteErrPacket(mysql.NewErr(mysql.ErrAbortingConnection, "should not go here", nil))
 		}
+		if _, err = clientIO.UpgradeToServerTLS(frontendTLSConfig); err != nil {
+			return errors.WithStack(err)
+		}
+		pkt, isSSL, err = clientIO.ReadSSLRequestOrHandshakeResp()
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		if isSSL {
+			return errors.WithStack(errors.New("expect handshake resp"))
+		}
+	} else {
+		binary.LittleEndian.PutUint32(pkt, (frontendCapability | pnet.ClientSSL).Uint32())
+	}
+	if commonCaps := frontendCapability & requiredCapabilities; commonCaps != requiredCapabilities {
+		logger.Error("require frontend capabilities", zap.Stringer("common", commonCaps), zap.Stringer("required", requiredCapabilities))
+		return errors.Wrapf(ErrCapabilityNegotiation, "require %s", requiredCapabilities&^commonCaps)
 	}
+	commonCaps := frontendCapability & proxyCapability
+	if frontendCapability^commonCaps != 0 {
+		logger.Error("frontend send capabilities unsupported by proxy", zap.Stringer("common", commonCaps), zap.Stringer("frontend", frontendCapability^commonCaps), zap.Stringer("proxy", proxyCapability^commonCaps))
+		return errors.Wrapf(ErrCapabilityNegotiation, "%s", frontendCapability^commonCaps)
+	}
+	resp := pnet.ParseHandshakeResponse(pkt)
+	auth.capability = commonCaps.Uint32()
+	auth.user = resp.User
+	auth.dbname = resp.DB
+	auth.collation = resp.Collation
+	auth.attrs = resp.Attrs
+
+	// write proxy header
+	if proxyProtocol {
+		proxy := clientIO.Proxy()
+		if proxy == nil {
+			proxy = &pnet.Proxy{
+				SrcAddress: clientIO.RemoteAddr(),
+				DstAddress: backendIO.RemoteAddr(),
+				Version:    pnet.ProxyVersion2,
+				Command:    pnet.ProxyCommandProxy,
+			}
+		}
+		if err := backendIO.WriteProxyV2(proxy); err != nil {
+			return err
+		}
+	}
+
+	// read server initial handshake
+	_, backendCapabilityU, err := auth.readInitialHandshake(backendIO)
 	if err != nil {
-		return err
+		return errors.WithStack(err)
 	}
-	if serverCapability&mysql.ClientSSL == 0 {
+	backendCapability := pnet.Capability(backendCapabilityU)
+	if common := commonCaps & backendCapability; (commonCaps^common)&^pnet.ClientSSL != 0 {
+		logger.Info("backend does not support capabilities from client&proxy", zap.Stringer("common", common), zap.Stringer("server", commonCaps^common), zap.Stringer("backend", backendCapability^common))
+	}
```

Comment:
This is not right, considering `ClientPluginAuthLenencClientData` and other capabilities that are in default backend capabilities.

Reply:
I don't get it. If `ClientPluginAuthLenencClientData` is supported, it should be sent by the backend.